nixpkgs_review/nix.py

```diff
@@ -89,42 +159,68 @@ def _nix_eval_filter(json: Dict[str, Any]) -> List[Attr]:
     return list(attr_by_path.values()) + broken
 
 
-def nix_eval(attrs: Set[str], system: str) -> List[Attr]:
-    attr_json = NamedTemporaryFile(mode="w+", delete=False)
-    delete = True
-    try:
-        json.dump(list(attrs), attr_json)
-        eval_script = str(ROOT.joinpath("nix/evalAttrs.nix"))
-        attr_json.flush()
-        cmd = [
-            "nix",
-            "--experimental-features",
-            "nix-command",
-            "--system",
-            system,
-            "eval",
-            "--json",
-            "--impure",
-            "--expr",
-            f"(import {eval_script} {attr_json.name})",
-        ]
-
+def nix_eval(
+    attrs: Set[str], system: str, cache_directory: Optional[Path] = None
+) -> List[Attr]:
+    def _eval(attrs: List[str]) -> Dict[str, Any]:
+        attr_json = NamedTemporaryFile(mode="w+", delete=False)
+        delete = True
         try:
-            nix_eval = subprocess.run(
-                cmd, check=True, stdout=subprocess.PIPE, text=True
-            )
-        except subprocess.CalledProcessError:
-            warn(
-                f"{' '.join(cmd)} failed to run, {attr_json.name} was stored inspection"
-            )
-            delete = False
-            raise
+            json.dump(attrs, attr_json)
+            eval_script = str(ROOT.joinpath("nix/evalAttrs.nix"))
+            attr_json.flush()
+            cmd = [
+                "nix",
+                "--experimental-features",
+                "nix-command",
+                "eval",
+                "--json",
+                "--system",
+                system,
+                "--impure",
+                "--expr",
+                f"(import {eval_script} {attr_json.name})",
+            ]
+
+            try:
+                nix_eval = subprocess.run(
+                    cmd, check=True, stdout=subprocess.PIPE, text=True
+                )
+            except subprocess.CalledProcessError:
+                warn(
+                    f"{' '.join(cmd)} failed to run, {attr_json.name} was stored inspection"
+                )
+                delete = False
+                raise
+
+            return json.loads(nix_eval.stdout)
+        finally:
+            attr_json.close()
+            if delete:
+                os.unlink(attr_json.name)
+
+    #
+    # Split the evaluation into chunks of 4096 attrs at a time.
+    # This helps limit the memory usage, which can be a problem.
+    #
```

Comment:
Does this take longer if you don't mind the RAM usage?

Reply:
Not really, since the chunks are so big (how often do you have 4096 attrs anyways?).

Reply:
When rebuilding staging that is quite common.

Reply:
I don't think it really effects the time very much at all.